### PR TITLE
MSI connection set up for Azure Container Registry for DockerV2 task

### PR DESCRIPTION
**Pull request by amaswauw**
**Task name**: DockerV2

**Description**: Added MSI authentication as a possibility to the DockerV2 task. (Service Principal is still supported.) The DockerV2 task now supports both service principal and managed identity log in.

**Documentation changes required:** (Y/N) Y

**Added unit tests:** (Y/N) Y

**Attached related issue:** (Y/N) Y. https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/1973804

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: common-npm-packages/docker-common-v2/package.json [hunk 1]
```diff
@@ -19,12 +19,13 @@
...
+      "azure-pipelines-tasks-azure-arm-rest-v2": "2.208.0"
     },
     "devDependencies": {
         "typescript": "4.0.2"
     },
-    "scripts": {
-        "build": "cd ../build-scripts && npm install && cd ../docker-common-v2 && node make.js"
-    }
+  "scripts": {
+    "build": "node ../build-scripts/downloadArchive.js https://vstsagenttools.blob.core.windows.net/tools/7zip/1/7zip.zip ./tools && node make.js"
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/package.json:29]: Can you please educate me why this change was required ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 2]
```diff
@@ -1,6 +1,9 @@
 "use strict";
 
 import * as tl from "azure-pipelines-task-lib/task";
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:3]: Sort your imports

File: Tasks/DockerV2/docker.ts [hunk 3]
```diff
@@ -7,60 +7,64 @@ import ContainerConnection from "azure-pipelines-tasks-docker-common-v2/containe
 import { getDockerRegistryEndpointAuthenticationToken } from "azure-pipelines-tasks-docker-common-v2/registryauthenticationprovider/registryauthenticationtoken";
 
 tl.setResourcePath(path.join(__dirname, 'task.json'));
+let registryAuthenticationToken: RegistryAuthenticationToken;
```
Comment (dhashrath) [inline Tasks/DockerV2/docker.ts:10]: Curious about this reordering. Also remove blank line at line 11

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts [hunk 4]
```diff
@@ -55,13 +55,23 @@ export default class RegistryServerAuthenticationToken {
...
     let authToken: RegistryServerAuthenticationToken;
 
     if (registryType === "ACR") {
         const loginServer = tl.getEndpointAuthorizationParameter(endpointId, "loginServer", false).toLowerCase();;
-        authToken = new ACRAuthenticationTokenProvider(endpointId, loginServer).getAuthenticationToken();
+        let authType: string;
+        try {
+            authType = tl.getEndpointAuthorizationScheme(endpointId, false);
+        } catch {
+            authType = tl.getEndpointAuthorizationParameter(endpointId, "scheme", false);
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:68]: Can this throw ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts [hunk 5]
```diff
@@ -55,13 +55,23 @@ export default class RegistryServerAuthenticationToken {
...
+export async function getDockerRegistryEndpointAuthenticationToken(endpointId: string): Promise<RegistryServerAuthenticationToken> {
     var registryType = tl.getEndpointDataParameter(endpointId, "registrytype", true);
     let authToken: RegistryServerAuthenticationToken;
 
     if (registryType === "ACR") {
         const loginServer = tl.getEndpointAuthorizationParameter(endpointId, "loginServer", false).toLowerCase();;
-        authToken = new ACRAuthenticationTokenProvider(endpointId, loginServer).getAuthenticationToken();
+        let authType: string;
+        try {
+            authType = tl.getEndpointAuthorizationScheme(endpointId, false);
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:66]: When can this code throw ? Can you catch more specific exception ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts (hunk 4 above)
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:68]: Add additional traces here

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts [hunk 6]
```diff
@@ -55,13 +55,23 @@ export default class RegistryServerAuthenticationToken {
     }
 }
 
-export function getDockerRegistryEndpointAuthenticationToken(endpointId: string): RegistryServerAuthenticationToken {
+export async function getDockerRegistryEndpointAuthenticationToken(endpointId: string): Promise<RegistryServerAuthenticationToken> {
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:58]: Can you please cover this with some unit tests ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts [hunk 7]
```diff
@@ -55,13 +55,23 @@ export default class RegistryServerAuthenticationToken {
...
         const loginServer = tl.getEndpointAuthorizationParameter(endpointId, "loginServer", false).toLowerCase();;
-        authToken = new ACRAuthenticationTokenProvider(endpointId, loginServer).getAuthenticationToken();
+        let authType: string;
+        try {
+            authType = tl.getEndpointAuthorizationScheme(endpointId, false);
+        } catch {
+            authType = tl.getEndpointAuthorizationParameter(endpointId, "scheme", false);
+        }
+        if (authType === "ManagedServiceIdentity") {
+            authToken = await new ACRAuthenticationTokenProvider(endpointId, loginServer).getMSIAuthenticationToken(0, 0);
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:71]: Readability: What does (0,0) signifies here ? Can you please label the parameters or add code comments ?

Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:71]: ACRAuthenticationTokenProvider(endpointId, loginServer) seems to be common to both MSI / SPN cases. Can instantiation be moved outside of if/else block ?

From C# experience, I always prefer to keep creation and method calls in separate lines and not chain them for better line numbers in stack traces.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 8]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
 
         return null;
     }
+
+    private _getACRToken(AADToken: string, retryCount: number, timeToWait: number): Q.Promise<string> {
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:52]: Move private methods to the bottom of the file. public methods followed by private methods

Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:52]: Can this be static method ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 9]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+        let deferred = Q.defer<string>();
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:53]: Can this piece be written using async / await pattern which you have used elsewhere in this PR ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 10]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+                        let waitedTime = 2000 + timeToWait * 2;
+                        retryCount += 1;
+                        setTimeout(() => {
+                            deferred.resolve(this._getACRToken(AADToken, retryCount, waitedTime));
+                        }, waitedTime);
+                    }
+                    else {
+                        deferred.reject(tl.loc('CouldNotFetchAccessTokenforACRStatusCode', response.statusCode, response.statusMessage));
+                    }
+
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:81]: nit: blank line

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 11]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+            (error) => {
+                deferred.reject(error)
+            }
+        );
+        return deferred.promise;
+    }
+
+    public async getMSIAuthenticationToken(retryCount: number, timeToWait: number): Promise<RegistryAuthenticationToken> {
+        if (this.registryURL && this.endpointName) {
+            let azureResourceManagerCommon1 = new azureResourceManagerCommon.ApplicationTokenCredentials("", "tempDomain", "", "https://management.core.windows.net/", "tempAuthorityURL", "tempActiveDirectoryResourceId", false, "ManagedServiceIdentity", null, null, null, null, null)
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:96]: new azureResourceManagerCommon() ?

Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:96]: Can you please label the parameters ?

what is azureResourceManagerCommon1 here ? Can you please qualify with better name ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 12]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+            let aadtoken = await azureResourceManagerCommon1.getToken();
+            let acrToken = await this._getACRToken(aadtoken, retryCount, timeToWait);
+            return new RegistryAuthenticationToken("00000000-0000-0000-0000-000000000000", acrToken, this.registryURL, "ManagedIdentity@AzureRM", this.getXMetaSourceClient());
+        }
+        return null;
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:101]: When this code can get hit ? I would prefer to throw rather than causing null ref in caller

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 13]
```diff
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+                }
+            },
+            (error) => {
+                deferred.reject(error)
+            }
+        );
+        return deferred.promise;
+    }
+
+    public async getMSIAuthenticationToken(retryCount: number, timeToWait: number): Promise<RegistryAuthenticationToken> {
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:94]: This is kind of OCP and SRP violation. Have you considered creating a new TokenProvider for MSI and have a factory which returns SPN or MSI Token Provider based on Service connection Auth type ? 

I would like to evaluate if that requires significant rework here

**Review by dhashrath — CHANGES_REQUESTED** [review]

File: Tasks/DockerV2/docker.ts (hunk 3 above)
Comment (amaswauw) [reply Tasks/DockerV2/docker.ts:10]: Reordering was necessary here because of the nature of the MSI authentication and the getting of the ACR token from two API calls. The SPN connection is synchronous in nature as the only two parameters needed can be received in constant time (the serviceprincipalid and the serviceprincipalkey). With MSI authentication, the first API call would get the AAD token and the second would get the ACR token. In this way, if the code was not asynchronous, we would not be able to initialize the registryAuthenticationToken variable correctly. I also noticed that the rest of the code uses the registryAuthenticationToken variable, so I decided to initialize the variable such that the token could be retrieved and then used in future transactions. This does not change the SPN login as there is nothing there that would result in the code needing to wait on the registry authentication token needing to be provided.

File: Tasks/DockerV2/make.json [hunk 14]
```diff
@@ -1,10 +1,19 @@
...
-		"compile" : true
-	}],
+	"common": [
+		{
+			"module": "../Common/utility-common-v2",
+			"type": "node",
+			"dest": "./",
+			"compile": true
+		},
+		{
```
[the remark this answers is not in the document]
Comment (amaswauw) [reply Tasks/DockerV2/make.json:9]: Similar to the package.json (build script), this change was needed in order to use the local version of the docker-common-v2 code and getting that code required the following lines of code to be changed.

File: Tasks/DockerV2/package.json [hunk 15]
```diff
@@ -5,7 +5,7 @@
     "@types/q": "^1.5.0",
     "@types/uuid": "^8.3.0",
     "azure-pipelines-task-lib": "^3.0.6-preview.0",
-    "azure-pipelines-tasks-docker-common-v2": "2.0.6",
+    "azure-pipelines-tasks-docker-common-v2": "file:../../_build/Tasks/Common/azure-pipelines-tasks-docker-common-v2-2.201.0.tgz",
```
[the remark this answers is not in the document]
Comment (amaswauw) [reply Tasks/DockerV2/package.json:8]: The switch here was necessary as the previous version of docker-common-v2 was deployed to npm. The changes that I have made would not be able to be on npm yet, and thus I needed a local build of the package to test the functionality that I have added in my code. Thus, the switch here was to account for that.

File: common-npm-packages/docker-common-v2/package.json [hunk 16]
```diff
@@ -19,12 +19,13 @@
         "@types/q": "1.5.4",
         "del": "2.2.0",
         "q": "1.4.1",
-        "azure-pipelines-task-lib": "^3.1.0"
+      "azure-pipelines-task-lib": "^3.1.0",
+      "azure-pipelines-tasks-azure-arm-rest-v2": "2.208.0"
```
[the remark this answers is not in the document]
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/package.json:23]: I thought the version of azure-pipelines-tasks-azure-arm-rest-v2 I needed was 2.208.0. I can, however, use the carrot notation (^) such that if the azure-pipelines-tasks-azure-arm-rest-v2 was updated, the code that I have made would pick up the new, updated package.

File: common-npm-packages/docker-common-v2/make.json [hunk 17]
```diff
@@ -0,0 +1,11 @@
+{
```
[the remark this answers is not in the document]
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/make.json:1]: I tried to replicate the generation of the local package - similar to the utility-common-v2 - such that the docker-common-v2 could use my local code as opposed to a publicly deployed npm package - as previously done with the prior version of the code. In this way, I had to introduce this file so that the code that I had made code be zipped correctly for testing purposes inside of ADO.

File: package-lock.json [hunk 18]
```diff
@@ -1,8 +1,1710 @@
 {
   "name": "Agent.Tasks",
   "version": "0.6.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
   "requires": true,
+  "packages": {
+    "": {
```
[the remark this answers is not in the document]
Comment (amaswauw) [reply package-lock.json:7]: I am a little bit unsure as to why there were this many packages required now that were not required before.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts (hunk 4 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:68]: This portion of code can throw an error. When I tried to log the changes inside of the ADO pipelines, I couldn't find specifically what error was thrown, but this can throw an error if the scheme is not successfully retrieved.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts (hunk 11 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:96]: Azure Resource Manager Common declares a new class called ApplicationTokenCredentials. Inside of this class there is the getToken() method which would get the MSI token which I am using inside of this function.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts (hunk 12 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:101]: I tried to replicate the SPN login here, in that, if there is not enough information that is provided, the code will not work/run and return a null pointer, or in the updated case, throw an error.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts (hunk 13 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:94]: I don't think that would require significant rework. I could use the getToken() method inside of arm-common-v2 and use that as a baseline. Small amount of rework, but it could work better in our case.

File: Tasks/DockerV2/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (dhashrath) [reply Tasks/DockerV2/package.json:8]: Please revert it before merge

File: Tasks/DockerV2/make.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (dhashrath) [reply Tasks/DockerV2/make.json:9]: I assume this is for testing and you will revert this before merge ?

File: common-npm-packages/docker-common-v2/make.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (dhashrath) [reply common-npm-packages/docker-common-v2/make.json:1]: I assume this is for testing and you will revert this before merge ?

File: common-npm-packages/docker-common-v2/package.json (hunk 1 above)
Comment (dhashrath) [reply common-npm-packages/docker-common-v2/package.json:29]: I assume this is for testing and you will revert this before merge ? can you confirm ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts (hunk 4 above)
Comment (dhashrath) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:68]: Do you think this should be safe handled then and fallback to SPN ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts (hunk 6 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/registryauthenticationprovider/registryauthenticationtoken.ts:58]: Not quite sure I understand. The change here only affects managed identity which can only be tested on a VM. The current SPN has unit tests which are inside the L0 file of the DockerV2 task. These tests all call the docker.ts file which is where the getDockerRegistryEndpointAuthenticationToken method is used. If you would like me to add more unit tests, I can do so, but this function is already being tested there with each call of a Docker Registry.

File: common-npm-packages/docker-common-v2/package.json (hunk 1 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/package.json:29]: I will remove this before merge.

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 19]
```diff
@@ -1,8 +1,11 @@
 "use strict";
 
+import AuthenticationTokenProvider from "./authenticationtokenprovider";
+import { ApplicationTokenCredentials } from "azure-pipelines-tasks-azure-arm-rest-v2/azure-arm-common";
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:4]: For my understanding - When do you use import {ApplicationTokenCredentials} from "./" over import ApplicationTokenCredentials from "./" without braces?

A

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 20]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
     }
+
+    public async getToken(): Promise<RegistryAuthenticationToken> {
+        let authType: string;
+        // Will error out with an internal error if the parameter is not found. This error is determined inside of the
+        // tl.getEndpointAuthorizationScheme/tl.getEndpointAuthorizationParameter and cannot be caught here as it is a
+        // custom error.
+        try {
+            authType = tl.getEndpointAuthorizationScheme(this.endpointName, false);
+        } catch (error) {
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:58]: can you please trace the error here and below since you are silently eating up the exception ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 21]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+        }
+        try {
+            authType = tl.getEndpointAuthorizationParameter(this.endpointName, "scheme", false);
+        } catch (error) {
+        }
+        if (!authType) {
+            authType = "ServicePrincipal";
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:65]: who refers this "ServicePrincipal" ? I do not see any code referring this

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 22]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+        let authType: string;
+        // Will error out with an internal error if the parameter is not found. This error is determined inside of the
+        // tl.getEndpointAuthorizationScheme/tl.getEndpointAuthorizationParameter and cannot be caught here as it is a
+        // custom error.
+        try {
+            authType = tl.getEndpointAuthorizationScheme(this.endpointName, false);
+        } catch (error) {
+        }
+        try {
+            authType = tl.getEndpointAuthorizationParameter(this.endpointName, "scheme", false);
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:61]: Do you need to call this only if getEndpointAuthorizationScheme errors out ? Otherwise, this will overwrite the value even if the former call succeeds ?

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 23]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+    private async _getMSIAuthenticationToken(retryCount: number, timeToWait: number): Promise<RegistryAuthenticationToken> {
+        if (this.registryURL && this.endpointName) {
+            try {
+                let aadtoken = await ApplicationTokenCredentials.getMSIAuthorizationToken(
+                    retryCount, timeToWait, "https://management.core.windows.net/", null);
+                let acrToken = await ACRAuthenticationTokenProvider._getACRToken(aadtoken, this.endpointName, this.registryURL, retryCount, timeToWait);
+                return new RegistryAuthenticationToken(
+                    "00000000-0000-0000-0000-000000000000", acrToken, this.registryURL,
+                    "ManagedIdentity@AzureRM", this.getXMetaSourceClient());
+            } catch (error) {
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:126]: Plese trace the actual error

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 24]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+        };
+        webClient.sendRequest(webRequest).then(
+            (response: webClient.WebResponse) => {
+                if (response.statusCode === 200) {
+                    deferred.resolve(response.body.refresh_token);
+                }
+                else if (response.statusCode == 429 || response.statusCode == 500) {
+                    if (retryCount < retryLimit) {
+                        let waitedTime = 2000 + timeToWait * 2;
+                        retryCount += 1;
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:97]: Trace error and that you are retrying

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 25]
```diff
@@ -42,7 +45,88 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+                        setTimeout(() => {
+                            deferred.resolve(this._getACRToken(AADToken, endpointName, registryURL, retryCount, waitedTime));
+                        }, waitedTime);
+                    }
+                    else {
+                        deferred.reject(tl.loc('CouldNotFetchAccessTokenforACRStatusCode', response.statusCode, response.statusMessage));
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:103]: Is this defined in resources file ? I for some reason only seeing 'MSIFetchError' as defined (unless I am looking at wrong resource file)

File: common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 26]
```diff
@@ -1,8 +1,11 @@
 "use strict";
 
+import AuthenticationTokenProvider from "./authenticationtokenprovider";
+import { ApplicationTokenCredentials } from "azure-pipelines-tasks-azure-arm-rest-v2/azure-arm-common";
+import RegistryAuthenticationToken from "./registryauthenticationtoken";
 import * as tl from "azure-pipelines-task-lib/task";
-import RegistryAuthenticationToken from "./registryauthenticationtoken"
-import AuthenticationTokenProvider from "./authenticationtokenprovider"
+import * as Q from "q";
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts:7]: Any reason you can't use async / await pattern for new code and have to fallback to Q ?

File: common-npm-packages/docker-common-v2/package.json [hunk 27]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "azure-pipelines-tasks-docker-common-v2",
-    "version": "2.198.1",
+    "version": "2.201.0",
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/package.json:3]: 2.210 ?

File: common-npm-packages/docker-common-v2/make.js [hunk 28]
```diff
@@ -9,5 +9,6 @@ util.run(path.join(__dirname, 'node_modules/.bin/tsc') + ' --outDir ' + buildPat
 util.cp(path.join(__dirname, 'package.json'), buildPath);
 util.cp(path.join(__dirname, 'package-lock.json'), buildPath);
 util.cp(path.join(__dirname, 'module.json'), buildPath);
+util.cp('-r', 'tools', buildPath);
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/make.js:12]: Can you please educate me about this change ?

File: common-npm-packages/azure-arm-rest-v2/Strings/resources.resjson/en-US/resources.resjson [hunk 29]
```diff
@@ -152,13 +152,12 @@
   "loc.messages.MethodNotImplementedError": "Method '%s' is not implemented.",
   "loc.messages.ExpiredServicePrincipal": "Could not fetch access token for Azure. Verify if the Service Principal used is valid and not expired. For more information refer https://aka.ms/azureappservicedeploytsg",
   "loc.messages.CorrelationIdForARM": "Correlation ID from ARM api call response : %s",
-  "loc.messages.CantDownloadAccessProfile": "Cannot download access profile/kube config file for the cluster %s. Reason %s.",
```
Comment (dhashrath) [inline common-npm-packages/azure-arm-rest-v2/Strings/resources.resjson/en-US/resources.resjson:155]: I did not understand changes to these files ? What it has to do with your code changes ?

File: Tasks/DockerV2/package.json [hunk 30]
```diff
@@ -5,7 +5,7 @@
     "@types/q": "^1.5.0",
     "@types/uuid": "^8.3.0",
     "azure-pipelines-task-lib": "^3.0.6-preview.0",
-    "azure-pipelines-tasks-docker-common-v2": "2.0.6",
+    "azure-pipelines-tasks-docker-common-v2": "^2.198.1",
```
Comment (dhashrath) [inline Tasks/DockerV2/package.json:8]: Looks like this was previously pinned to some specific reason. Can you check the history why this was done such a way to be on safer side?

**Review by dhashrath — COMMENTED** [review]
Thanks Aman for resolving most of the feedback. Please take a look at pending feedback and see if they make sense.

I would request you to add more traces so if something regresses here, on call engineer can enable traces to find out and debug easily when you go back to continue your education.

Secondly, think about some useful telemetry that you can add here to measure business impact. You can make these changes in separate PR.

File: common-npm-packages/docker-common-v2/package.json (hunk 27 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/package.json:3]: Missed this when I was bumping up versions everywhere else. This has now been changed to the correct version of 2.210.0

File: common-npm-packages/azure-arm-rest-v2/Strings/resources.resjson/en-US/resources.resjson (hunk 29 above)
Comment (amaswauw) [reply common-npm-packages/azure-arm-rest-v2/Strings/resources.resjson/en-US/resources.resjson:155]: Not sure why these lines were deleted and the others were added. I did not manually delete or add any of these lines, they were added when I built the project and connected it to docker-common-v2. I can revert these changes if need be.

File: Tasks/DockerV2/package.json (hunk 30 above)
Comment (amaswauw) [reply Tasks/DockerV2/package.json:8]: Looks like the last time this file was updated was over a year ago, in which the docker-common-v2 file was updated from 2.0.4 to 2.0.6. I can revert this back to 2.0.6 if need be, but we may need to update this file in the future as 2.0.6 is slightly older than the current version of docker-common-v2 that is currently supported.

File: common-npm-packages/docker-common-v2/make.js (hunk 28 above)
Comment (amaswauw) [reply common-npm-packages/docker-common-v2/make.js:12]: This was a mistake on my part. This line was needed for testing purposes, and I thought I reverted it back to the original version of the file, but it looks like I didn't. I have gone ahead and removed this line from the file.

File: common-npm-packages/docker-common-v2/package.json [hunk 31]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "azure-pipelines-tasks-docker-common-v2",
-    "version": "2.198.1",
+    "version": "2.210.0",
```
Comment (dhashrath) [inline common-npm-packages/docker-common-v2/package.json:3]: Bump to 211 ?

**Review by dhashrath — APPROVED** [review]
Please include details of all the testing you have done in your PR description and also link the wiki.

**Comment by amaswauw** [comment]
Testing: I have tested these changes inside of a mwittest organization. These tests prove that the current functionality of the ServicePrincipal authentication (the current scheme used by Azure DevOps surrounding the DockerV2 task and ACRs still work to this day. The new changes will be an additional layer of a new authentication scheme, and will not change any of the original functionality, but it would rather add additional support in terms of a new authentication scheme, being managed service identity.

The wiki that I am authoring is this one right here: https://dev.azure.com/mseng/AzureDevOps/_wiki/wikis/AzureDevOps.wiki/30608/Engineering-Environment-Setup-DockerV2-Task
This wiki goes into depth about the current changes that I have made to the system (not including the patch bumps to the tasks that are changed due to the docker-common-v2 being updated.

**Review by mmrazik — APPROVED** [review]
